src/core/internal/hash.d

```diff
@@ -30,8 +30,20 @@ private enum floatCoalesceNaNs = true;
         assert(hashOf(double.nan) == hashOf(-double.nan)); // Same hash for different NaN.
 }
 
+private enum hasCallableToHash(T) = __traits(compiles,
+    {
+        size_t x = ((T* x) => (*x).toHash())(null);
+    });
```

Comment:
This PR also contains a relaxation of the required signature of `toHash` for structs & unions. If the type implicitly converts to `size_t` it's fine. This originally was in https://github.com/dlang/druntime/pull/2240 but that PR has not been pulled yet.